services/device/geolocation/geoclue_location_provider.cc

```diff
@@ -0,0 +1,180 @@
+
+namespace {
+
+constexpr char kBraveDesktopId[] = "com.brave.Browser";
+
+// Note: This method blocks because the call to NewSystemProvider is not
+// asynchronous, but it happens on the background geolocation thread.
+//
+// The easiest way to determine if a DBus service exists is to try and call a
+// method on it, and see if it fails. For this, I chose
+// |GeoClue2.Manager.GetClient| (this is cached on the GeoClue2 side, so it will
+// be the same client we get when we start our service).
+bool GeoClueAvailable() {
+  base::ScopedBlockingCall scoped_blocking_call(FROM_HERE,
+                                                base::BlockingType::WILL_BLOCK);
+
+  dbus::Bus::Options options;
+  options.bus_type = dbus::Bus::SYSTEM;
+  options.connection_type = dbus::Bus::PRIVATE;
+  auto bus = base::MakeRefCounted<dbus::Bus>(options);
+
+  dbus::ObjectProxy* proxy = bus->GetObjectProxy(
+      GeoClueClientObject::kServiceName,
+      dbus::ObjectPath(GeoClueClientObject::kManagerObjectPath));
+
+  dbus::MethodCall call(GeoClueClientObject::kManagerInterfaceName,
+                        "GetClient");
+  auto response =
+      proxy->CallMethodAndBlock(&call, dbus::ObjectProxy::TIMEOUT_USE_DEFAULT);
+
+  // If the response is |nullptr| then the GeoClue2.Manager does not exist.
+  bool available = response.get();
+
+  // Shutdown this bus - we'll use one on the DBus thread for our actual
+  // provider.
+  bus->ShutdownAndBlock();
+
+  return available;
+}
+
+}  // namespace
+
+GeoClueLocationProvider::GeoClueLocationProvider() {
+  dbus::Bus::Options options;
+  options.bus_type = dbus::Bus::SYSTEM;
+  options.connection_type = dbus::Bus::PRIVATE;
+  options.dbus_task_runner = dbus_thread_linux::GetTaskRunner();
+
+  bus_ = base::MakeRefCounted<dbus::Bus>(options);
+}
+
+GeoClueLocationProvider::~GeoClueLocationProvider() {
+  dbus_thread_linux::GetTaskRunner()->PostTask(
+      FROM_HERE, base::BindOnce(&dbus::Bus::ShutdownAndBlock, std::move(bus_)));
+}
+
+void GeoClueLocationProvider::SetUpdateCallback(
+    const LocationProviderUpdateCallback& callback) {
+  position_update_callback_ = callback;
+}
+
+void GeoClueLocationProvider::StartProvider(bool high_accuracy) {
+  // The GeoClue2 Client doesn't provide a location with the updated accuracy
+  // unless it is restarted, so if the accuracy level has changed we need to
+  // restart the provider.
+  if (!high_accuracy_requested_ && high_accuracy) {
+    client_.reset();
+    high_accuracy_requested_ = true;
+  }
+
+  if (!client_) {
+    GeoClueClientObject::CreateParams params;
+    params.bus = bus_;
+    params.desktop_id = kBraveDesktopId;
+    params.high_accuracy = high_accuracy_requested_;
+    params.on_location_changed = base::BindRepeating(
+        &GeoClueLocationProvider::OnLocationUpdated, base::Unretained(this));
+    params.on_error = base::BindRepeating(&GeoClueLocationProvider::OnError,
+                                          base::Unretained(this));
```

Comment:
`Unretained` is fine here, because `GeoClueClientObject` is owned by `GeoClueLocationProvider` and these callbacks are called from `GeoClueClientObject` members via `WeakPtr`.